pageserver/src/http/routes.rs

```diff
@@ -1091,6 +1175,64 @@ async fn post_tracing_event_handler(mut r: Request<Body>) -> Result<Response<Bod
     json_response(StatusCode::OK, ())
 }
 
+/// Common functionality of all the HTTP API handlers.
+///
+/// - Adds a tracing span to each request (by `request_span`)
+/// - Logs the request depending on the request method (by `request_span`)
+/// - Logs the response if it was not successful (by `request_span`
+/// - Shields the handler function from async cancellations. Hyper can drop the handler
+///   Future if the connection to the client is lost, but most of the pageserver code is
+///   not async cancellation safe. This converts the dropped future into a graceful cancellation
+///   request with a CancellationToken.
+async fn api_handler<R, H>(request: Request<Body>, handler: H) -> Result<Response<Body>, ApiError>
+where
+    R: std::future::Future<Output = Result<Response<Body>, ApiError>> + Send + 'static,
+    H: FnOnce(Request<Body>, CancellationToken) -> R + Send + Sync + 'static,
+{
+    // Spawn a new task to handle the request, to protect the handler from unexpected
+    // async cancellations. Most pageserver functions are not async cancellation safe.
+    // We arm a drop-guard, so that if Hyper drops the Future, we signal the task
+    // with the cancellation token.
+    let token = CancellationToken::new();
+    let guard = token.clone().drop_guard();
+    let handle = tokio::spawn(request_span(request, move |r| handler(r, token)));
+
+    let result = handle.await;
+
+    guard.disarm();
+
+    match result {
+        Ok(result) => result,
+        Err(e) => {
+            error!("HTTP request handler task panicked: {e:?}");
+            Err(ApiError::InternalServerError(anyhow!(
+                "HTTP request handler task panicked"
+            )))
+        }
+    }
```

Comment:
At first, I wrote this "cancellation shield" as a separate, generic, library function. It looked like this:
```
/// Shield an async Future from cancellation.
pub async fn cancel_shield<M, F>(make_fut: M) -> F::Output
where
    F: Future + Send + 'static,
    F::Output: Send,
    M: FnOnce(CancellationToken) -> F
{
    let token = CancellationToken::new();
    let guard = token.clone().drop_guard();
    let handle = tokio::spawn(make_fut(token));

    let result = handle.await;

    guard.disarm();
    match result {
        Ok(result) => result,
        Err(e) => {
            panic!("joinerror: {e}");
        }
    }
}
```
It would be kind of nice to have a generic helper like that, but in the end I decided not to. Reasons:
- This is very short and straightforward, so it's not a lot of repetition to just open-code it everywhere
- There was only caller of this currently
- It would be nice to turn the JoinError into an Error rather than panic, but  that would require a more specialized error type, which would make this less generic, and would require some more boilerplate in the callers again.